### ISSUE TITLE
Use consistency_fail gem to add db indexes for uniqueness constraints

### ISSUE THREAD

**Issue by cofiem** [QutEcoacoustics/baw-server#138]
https://github.com/trptcolin/consistency_fail

```
There are calls to validates_uniqueness_of that aren't backed by unique indexes.
--------------------------------------------------------------------------------
Model           Table Columns
--------------------------------------------------------------------------------
AudioRecording  audio_recordings (uuid)
AudioRecording  audio_recordings (file_hash)
Bookmark        bookmarks (name, creator_id)
Dataset         datasets (name, creator_id)
Job             jobs (name)
Permission      permissions (level, user_id, project_id)
Permission      permissions (level, logged_in_user, project_id)
Permission      permissions (level, anonymous_user, project_id)
Project         projects (name)
Tag             tags (text)
Tagging         audio_events_tags (audio_event_id, tag_id, audio_event_id)
Tagging         audio_events_tags (tag_id, tag_id, audio_event_id)
--------------------------------------------------------------------------------


There are calls to has_one that aren't backed by unique indexes.
--------------------------------------------------------------------------------
Model   Table Columns
--------------------------------------------------------------------------------
Script  scripts (updated_by_script_id)
Script  scripts (original_script_id)
--------------------------------------------------------------------------------

Hooray! All calls to has_one_with_polymorphic are correctly backed by a unique index.
```

**Comment by cofiem** [comment]
We need to make decisions on what about this.

````
    add_index  :bookmarks, [:name, :creator_id], name: 'bookmarks_name_creator_id_uidx', unique: true, order: {name: :asc}
    add_index  :datasets, [:name, :creator_id], name: 'datasets_name_creator_id_uidx', unique: true, order: {name: :asc}

    add_index  :jobs, :name, name: 'jobs_name_uidx', unique: true, order: {name: :asc}

    add_index  :permissions, [:project_id, :level, :user_id], name: 'permissions_level_user_id_project_id_uidx', unique: true, order: {project_id: :asc, user_id: :asc}

    add_index  :projects, :name, name: 'projects_name_uidx', unique: true, order: {name: :asc}

    add_index  :tags, :text, name: 'tags_text_uidx', unique: true, order: {text: :asc}

    add_index  :scripts, :updated_by_script_id, name: 'scripts_updated_by_script_id_uidx', unique: true

    add_index  :audio_recordings, :uuid, name: 'audio_recordings_uuid_uidx', unique: true

    # requires modifying the file_hash before this index will succeed
    #execute "UPDATE audio_recordings SET file_hash = null WHERE file_hash = 'SHA256::'"

    add_index  :audio_recordings, :file_hash, name: 'audio_recordings_file_hash_uidx', unique: true, where: "file_hash <> 'SHA256::'"
    # PG::UniqueViolation: ERROR:  could not create unique index "audio_recordings_file_hash_uidx"
    # DETAIL:  Key (file_hash)=(SHA256::) is duplicated.
    #: CREATE UNIQUE INDEX  "audio_recordings_file_hash_uidx" ON "audio_recordings"  ("file_hash")
````

**Comment by cofiem** [comment]
Also add indexes on the foreign keys (yes, apparently we don't have these. What. The. Heck.)

````
add_index :audio_event_comments, :audio_event_id
add_index :audio_event_comments, :creator_id
add_index :audio_event_comments, :updater_id
add_index :audio_event_comments, :deleter_id
add_index :audio_event_comments, :flagger_id
add_index :audio_recordings, :creator_id
add_index :audio_recordings, :updater_id
add_index :audio_recordings, :deleter_id
add_index :audio_recordings, :uploader_id
add_index :audio_events, :audio_recording_id
add_index :audio_events, :creator_id
add_index :audio_events, :updater_id
add_index :audio_events, :deleter_id
add_index :bookmarks, :audio_recording_id
add_index :bookmarks, :creator_id
add_index :bookmarks, :updater_id
add_index :projects, :creator_id
add_index :projects, :updater_id
add_index :projects, :deleter_id
add_index :permissions, [:project_id, :user_id]
add_index :permissions, :project_id
add_index :permissions, :user_id
add_index :permissions, :creator_id
add_index :permissions, :updater_id
add_index :projects_sites, [:project_id, :site_id]
add_index :projects_sites, :left_side_id
add_index :projects_sites, :site_id
add_index :projects_sites, :project_id
add_index :projects_saved_searches, [:project_id, :saved_search_id]
add_index :projects_saved_searches, :left_side_id
add_index :projects_saved_searches, :saved_search_id
add_index :projects_saved_searches, :project_id
add_index :analysis_jobs, :creator_id
add_index :analysis_jobs, :updater_id
add_index :analysis_jobs, :deleter_id
add_index :analysis_jobs, :script_id
add_index :analysis_jobs, :saved_search_id
add_index :saved_searches, :creator_id
add_index :saved_searches, :deleter_id
add_index :scripts, :creator_id
add_index :scripts, :updated_by_script_id
add_index :sites, :creator_id
add_index :sites, :updater_id
add_index :sites, :deleter_id
add_index :tags, :creator_id
add_index :tags, :updater_id
add_index :audio_events_tags, :creator_id
add_index :audio_events_tags, :updater_id
````

**Comment by cofiem** [comment]
https://github.com/plentz/lol_dba

**Comment by atruskie** [comment]
:+1:

**Comment by atruskie** [comment]
I believe this was closed in 215cc13
